x/mongo/driver/crypt.go

```diff
@@ -382,3 +389,69 @@ func (c *crypt) decryptKey(kmsCtx *mongocrypt.KmsContext) error {
 		}
 	}
 }
+
+// needsKmsProvider returns true if provider was initially set to an empty document.
+// An empty document signals the driver to fetch credentials.
+func needsKmsProvider(kmsProviders *bsoncore.Document, provider string) bool {
+	gcpVal, err := kmsProviders.LookupErr(provider)
+	if err != nil {
+		// KMS provider is not configured.
+		return false
+	}
+	gcpDoc, ok := gcpVal.DocumentOK()
+	if !ok || len(gcpDoc) != 5 {
+		// KMS provider is not an empty document.
+		return false
+	}
+	return true
+}
+
+func (c *crypt) provideKmsProviders(cryptCtx *mongocrypt.Context) error {
+	kmsProviders := c.mongoCrypt.GetKmsProviders()
+	builder := bsoncore.NewDocumentBuilder()
+
+	if needsKmsProvider(kmsProviders, "gcp") {
+		// "gcp" KMS provider is an empty document.
+		// Attempt to fetch from GCP Instance Metadata server.
+		{
+			metadataHost := "metadata.google.internal"
+			if os.Getenv("GCE_METADATA_HOST") != "" {
+				metadataHost = os.Getenv("GCE_METADATA_HOST")
+			}
+			url := fmt.Sprintf("http://%s/computeMetadata/v1/instance/service-accounts/default/token", metadataHost)
+			req, err := http.NewRequest("GET", url, nil)
+			if err != nil {
+				return internal.WrapErrorf(err, "unable to retrieve GCP credentials")
+			}
+			req.Header.Set("Metadata-Flavor", "Google")
+			resp, err := http.DefaultClient.Do(req.WithContext(context.Background()))
+			if err != nil {
+				return internal.WrapErrorf(err, "unable to retrieve GCP credentials")
+			}
+			defer func() { _ = resp.Body.Close() }()
```

Comment:
Optional: The linter rules allow ignoring errors returned by `Close()` functions.

E.g.
```go
defer resp.Body.Close()
```

Reply:
I think it's actually much safer to do it the way @matthewdale is suggesting. Go will only evaluate functions on the first level of defer, [here](https://go.dev/play/p/FoBhNFkGC8t) is an example where this breaks. [This is also the idiomatic way of closing the response body](https://pkg.go.dev/net/http#:~:text=The%20client%20must%20close%20the%20response%20body%20when%20finished%20with%20it%3A).